File: Application/UseCases/ExportInterviewUseCase.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;
using InterviewMasterApp.Domain.Model;

namespace InterviewMasterApp.Application.UseCases
{
    /// <summary>
    /// Enum for export formats.
    /// Migrated from Swift: ExportInterviewUseCase.swift
    /// </summary>
    public enum ExportFormat
    {
        Markdown,
        Json
    }

    /// <summary>
    /// Data structure for exported interview data (JSON serializable).
    /// </summary>
    public class ExportData
    {
        public string Version { get; set; }
        public string ExportDate { get; set; }
        public ExportSettings Settings { get; set; }
        public List<ExportMessage> Messages { get; set; }

        public class ExportSettings
        {
            public string Role { get; set; }
            public string ProgrammingLanguage { get; set; }
            public string SpeakingLanguage { get; set; }
            public string Frameworks { get; set; }
        }

        public class ExportMessage
        {
            public string Type { get; set; }
            public string Content { get; set; }
            public string Topic { get; set; }
            public string Timestamp { get; set; }
            public string AudioSource { get; set; }
            public int? LatencyMs { get; set; }
        }
    }

    /// <summary>
    /// Use case for exporting interview transcript to Markdown or JSON format.
    /// Migrated from Swift: ExportInterviewUseCase.swift
    /// </summary>
    public class ExportInterviewUseCase
    {
        /// <summary>
        /// Execute export operation.
        /// </summary>
        /// <param name="messages">Interview messages to export</param>
        /// <param name="format">Export format (Markdown or JSON)</param>
        /// <returns>Formatted export string</returns>
        public string Execute(List<InterviewMessage> messages, ExportFormat format)
        {
            // Filter for exportable message types
            var exportable = messages
                .Where(msg => msg.Type is InterviewMessage.MessageType.Question
                           or InterviewMessage.MessageType.Answer
                           or InterviewMessage.MessageType.FollowUp
                           or InterviewMessage.MessageType.CodingTask)
                .ToList();

            return format switch
            {
                ExportFormat.Markdown => FormatMarkdown(exportable),
                ExportFormat.Json => FormatJson(exportable),
                _ => "{}"
            };
        }

        /// <summary>
        /// Format messages as Markdown.
        /// </summary>
        private static string FormatMarkdown(List<InterviewMessage> messages)
        {
            var settings = AppSettings.Shared; // or AppSettings.Shared equivalent
            var sb = new StringBuilder();

            sb.AppendLine("# Interview Transcript\n");
            sb.AppendLine($"**Date:** {DateTime.Now:G}");
            sb.AppendLine($"**Role:** {settings.RoleDisplayName}");
            sb.AppendLine($"**Language:** {settings.LanguageDisplayName}");
            sb.AppendLine($"**Response Language:** {settings.ResponseLanguageDisplayName}");

            if (!string.IsNullOrEmpty(settings.Frameworks))
            {
                sb.AppendLine($"**Frameworks:** {settings.Frameworks}");
            }

            sb.AppendLine("\n**Legend:**");
            sb.AppendLine("- 🎙️ **Interviewer** - Questions from the interviewer");
            sb.AppendLine("- 💡 **Suggested Answer** - AI-generated answer hints\n");
            sb.AppendLine("---\n");

            string currentTopic = null;

            foreach (var msg in messages)
            {
                // Add topic header if changed
                if (!string.IsNullOrEmpty(msg.Topic) && msg.Topic != currentTopic && msg.Topic != "unknown")
                {
                    sb.AppendLine($"## {CapitalizeFirstLetter(msg.Topic)}\n");
                    currentTopic = msg.Topic;
                }

                var time = msg.DisplayTime;

                switch (msg.Type)
                {
                    case InterviewMessage.MessageType.Question:
                        sb.AppendLine($"### 🎙️ Interviewer <small>({time})</small>\n");
                        sb.AppendLine($"> {msg.Content}\n");
                        break;

                    case InterviewMessage.MessageType.Answer:
                    case InterviewMessage.MessageType.FollowUp:
                        var header = "### 💡 Suggested Answer";
                        if (msg.DisplayLatency != null)
                        {
                            header += $" <small>({msg.DisplayLatency})</small>";
                        }
                        sb.AppendLine(header + "\n");
                        sb.AppendLine($"{msg.Content}\n");
                        break;

                    case InterviewMessage.MessageType.CodingTask:
                        sb.AppendLine("### 💻 Coding Solution\n");
                        sb.AppendLine($"{msg.Content}\n");
                        break;
                }
            }

            sb.AppendLine("---\n");
            sb.AppendLine("*Exported from Interview Master*");

            return sb.ToString();
        }

        /// <summary>
        /// Format messages as JSON.
        /// </summary>
        private static string FormatJson(List<InterviewMessage> messages)
        {
            var settings = AppSettings.Shared;
            var exportMessages = messages.Select(msg => new ExportData.ExportMessage
            {
                Type = GetTypeString(msg.Type),
                Content = msg.Content,
                Topic = msg.Topic,
                Timestamp = msg.Timestamp.ToUniversalTime().ToString("O"),
                AudioSource = msg.AudioSource?.ToString(),
                LatencyMs = msg.ResponseLatencyMs
            }).ToList();

            var data = new ExportData
            {
                Version = "1.0",
                ExportDate = DateTime.UtcNow.ToString("O"),
                Settings = new ExportData.ExportSettings
                {
                    Role = settings.RoleDisplayName,
                    ProgrammingLanguage = settings.LanguageDisplayName,
                    SpeakingLanguage = settings.ResponseLanguageDisplayName,
                    Frameworks = settings.Frameworks
                },
                Messages = exportMessages
            };

            var options = new JsonSerializerOptions
            {
                WriteIndented = true,
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase
            };

            return JsonSerializer.Serialize(data, options);
        }

        /// <summary>
        /// Get string representation of message type.
        /// </summary>
        private static string GetTypeString(InterviewMessage.MessageType type)
        {
            return type switch
            {
                InterviewMessage.MessageType.Question => "question",
                InterviewMessage.MessageType.Answer => "answer",
                InterviewMessage.MessageType.FollowUp => "followUp",
                InterviewMessage.MessageType.CodingTask => "codingTask",
                InterviewMessage.MessageType.UserResponse => "userResponse",
                InterviewMessage.MessageType.Status => "status",
                InterviewMessage.MessageType.Screenshot => "screenshot",
                _ => "unknown"
            };
        }

        /// <summary>
        /// Capitalize first letter of string.
        /// </summary>
        private static string CapitalizeFirstLetter(string text)
        {
            if (string.IsNullOrEmpty(text))
                return text;

            return char.ToUpper(text[0]) + text.Substring(1);
        }
    }
}

File: Application/UseCases/ExportInterviewUseCase.swift
import Foundation

enum ExportFormat: String {
    case markdown
    case json
}

struct ExportData: Codable {
    let version: String
    let exportDate: String
    let settings: ExportSettings
    let messages: [ExportMessage]

    struct ExportSettings: Codable {
        let role: String
        let programmingLanguage: String
        let listeningLanguage: String
        let responseLanguage: String
        let speakingLanguage: String
        let frameworks: String
    }

    struct ExportMessage: Codable {
        let type: String
        let content: String
        let topic: String?
        let timestamp: String
        let audioSource: String?
        let latencyMs: Int?
    }
}

struct ExportInterviewUseCase {

    func execute(messages: [InterviewMessage], format: ExportFormat) -> String {
        let exportable = messages.filter { msg in
            switch msg.type {
            case .question, .answer, .followUp, .codingTask:
                return true
            default:
                return false
            }
        }

        switch format {
        case .markdown:
            return formatMarkdown(exportable)
        case .json:
            return formatJSON(exportable)
        }
    }

    private func formatMarkdown(_ messages: [InterviewMessage]) -> String {
        let settings = AppSettings.shared

        var md = "# Interview Transcript\n\n"
        md += "**Date:** \(DateFormatter.localizedString(from: Date(), dateStyle: .long, timeStyle: .short))\n"
        md += "**Role:** \(settings.role.displayName)\n"
        md += "**Programming Language:** \(settings.programmingLanguage.displayName)\n"
        md += "**Listening Language:** \(settings.listeningLanguage.displayName)\n"
        md += "**Response Language:** \(settings.responseLanguage.displayName)\n"
        if !settings.frameworks.isEmpty {
            md += "**Frameworks:** \(settings.frameworks)\n"
        }
        md += "\n"
        md += "**Legend:**\n"
        md += "- 🎙️ **Interviewer** - Questions from the interviewer\n"
        md += "- 💡 **Suggested Answer** - AI-generated answer hints\n\n"
        md += "---\n\n"

        var currentTopic: String? = nil

        for msg in messages {
            if let topic = msg.topic, topic != currentTopic, topic != "unknown" {
                md += "## \(topic.capitalized)\n\n"
                currentTopic = topic
            }

            let time = msg.displayTime
            switch msg.type {
            case .question:
                md += "### 🎙️ Interviewer <small>(\(time))</small>\n\n"
                md += "> \(msg.content)\n\n"
            case .answer, .followUp:
                var header = "### 💡 Suggested Answer"
                if let latency = msg.displayLatency {
                    header += " <small>(\(latency))</small>"
                }
                md += "\(header)\n\n"
                md += "\(msg.content)\n\n"
            case .codingTask:
                md += "### 💻 Coding Solution\n\n"
                md += "\(msg.content)\n\n"
            default:
                break
            }
        }

        md += "---\n\n*Exported from Interview Master*\n"
        return md
    }

    private func formatJSON(_ messages: [InterviewMessage]) -> String {
        let settings = AppSettings.shared
        let dateFormatter = ISO8601DateFormatter()

        let exportMessages = messages.map { msg in
            ExportData.ExportMessage(
                type: typeString(msg.type),
                content: msg.content,
                topic: msg.topic,
                timestamp: dateFormatter.string(from: msg.timestamp),
                audioSource: msg.audioSource.map { $0 == .microphone ? "microphone" : "system" },
                latencyMs: msg.responseLatencyMs
            )
        }

        let data = ExportData(
            version: "1.0",
            exportDate: dateFormatter.string(from: Date()),
            settings: ExportData.ExportSettings(
                role: settings.role.displayName,
                programmingLanguage: settings.programmingLanguage.displayName,
                listeningLanguage: settings.listeningLanguage.displayName,
                responseLanguage: settings.responseLanguage.displayName,
                speakingLanguage: settings.responseLanguage.displayName,
                frameworks: settings.frameworks
            ),
            messages: exportMessages
        )

        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]

        guard let jsonData = try? encoder.encode(data),
              let jsonString = String(data: jsonData, encoding: .utf8) else {
            return "{}"
        }

        return jsonString
    }

    private func typeString(_ type: InterviewMessage.MessageType) -> String {
        switch type {
        case .question: return "question"
        case .answer: return "answer"
        case .followUp: return "followUp"
        case .codingTask: return "codingTask"
        case .userResponse: return "userResponse"
        case .status: return "status"
        case .screenshot: return "screenshot"
        }
    }
}
